src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.EasyRequest.cs

```diff
@@ -249,10 +249,51 @@ private void SetUrl()
                 }
 
                 EventSourceTrace("Url: {0}", requestUri);
-                SetCurlOption(CURLoption.CURLOPT_URL, requestUri.AbsoluteUri);
+                if (requestUri.Host != requestUri.IdnHost)
+                {
+                    SetCurlOption(CURLoption.CURLOPT_URL, ReplaceHostWithIdnHost(requestUri.AbsoluteUri, requestUri.Host, requestUri.IdnHost));
+                }
+                else
+                {
+                    SetCurlOption(CURLoption.CURLOPT_URL, requestUri.AbsoluteUri);
+                }
+
                 SetCurlOption(CURLoption.CURLOPT_PROTOCOLS, (long)(CurlProtocols.CURLPROTO_HTTP | CurlProtocols.CURLPROTO_HTTPS));
             }
 
+            private unsafe string ReplaceHostWithIdnHost(string absoluteUri, string host, string idnHost)
+            {
+                int start = absoluteUri.IndexOf(host);
```

Comment:
In URI, the authority component is defined as 

```
user:password@host:port
```

So yes, if the user or password component has host string in it, the wrong part will be replaced. The implementation can be changed to account to parse this syntax for getting the first index of host string that needs to be replaced. Nevertheless, testing UriBuilder seems to replace the correct part, tested with this test case:

```c#
Uri uri = new Uri("http://allehjørnet:allehjørnet@allehjørnet.no:123");
Console.WriteLine(new UriBuilder(uri) { Host = uri.IdnHost }.Uri.AbsoluteUri);
```

outputs,

```
http://allehj%C3%B8rnet:allehj%C3%B8rnet@xn--allehjrnet-5cb.no:123/
```